### PR TITLE
Retrieve Trouter Config From ChatGW

### PR DESCRIPTION
**Pull request by angellan-msft**
This PR is to add the logic to retrieve TrouterServiceUrl and RegistrarServiceUrl from Chat Gateway rather than hardcode them on SDK side. 
Design doc: https://microsoft.sharepoint-df.com/:w:/t/IC3SDK-Chat-Private/EQMKswlC5SdPnYQdgUpcDncBM81lDulf6I5xPpCyy7AyuA?e=PcjUx6

**How to test?**
I used the sample app for E2E tests and confirmed it can work well.
It's difficult to add unit tests for this change.(Previously we didn't have tests for SignalingClient as well).

### CODE REVIEW

File: sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift [hunk 1]
```diff
@@ -40,56 +40,93 @@ public typealias CommunicationSignalingErrorHandler = (CommunicationSignalingErr
...
+                let trouterSkypeTokenHeaderProvider = TrouterSkypetokenAuthHeaderProvider(
+                    skypetokenProvider: self.communicationSkypeTokenProvider
+                )
+                
+                // Remove the "https://" prefix from the URLs
+                let trouterHostname = configuration.trouterServiceUrl.replacingOccurrences(of: "https://", with: "")
+                let registrarUrlWithoutScheme = configuration.registrarServiceUrl.replacingOccurrences(of: "https://", with: "")
+                
+                // Remove the "v3/registrations" or "V3/registrations" suffix from the regsitrar URL
+                let registrarBasePath = registrarUrlWithoutScheme
```
Comment (LuChen-Microsoft) [inline sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift:76]: Please ask Trouter team to understand more about the Trouter service URL

File: sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift [hunk 2]
```diff
@@ -98,18 +135,73 @@ class CommunicationSignalingClient {
...
             communicationHandlers.removeValue(forKey: event)
         }
     }
+    
+    private func getTrouterSettings(
+            token: String,
+            endpoint: String,
+            completionHandler: @escaping (Result<RealTimeNotificationConfiguration, AzureError>) -> Void
+        ) {
+            let urlString = "\(endpoint)/chat/config/realTimeNotifications?api-version=2024-09-01"
```
Comment (LuChen-Microsoft) [inline sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift:168]: Please put the api version into a separate constant

File: sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift [hunk 3]
```diff
@@ -98,18 +135,73 @@ class CommunicationSignalingClient {
...
+                return
+            }
+
+            var request = URLRequest(url: url)
+            request.httpMethod = "GET"
+            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
+
+            let session = URLSession.shared
+            session.dataTask(with: request) { data, response, error in
+                if let error = error {
```
Comment (LuChen-Microsoft) [inline sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift:180]: Please throw communication error

File: sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift (hunk 2 above)
Comment (angellan-msft) [reply sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift]: Put it in the constant "configApiVersion" right now.

File: sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift (hunk 3 above)
Comment (angellan-msft) [reply sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift:181]: Yes we throw AzureError here, which is the type of error thrown in other places within the project.

File: sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift (hunk 1 above)
Comment (angellan-msft) [reply sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift]: Based on Jiri's suggestion, we should remove the "/v3/a" suffixes from both Trouter endpoint and Registrar endpoint on service side and let client side to append the values. 

Updated.

File: sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift [hunk 4]
```diff
@@ -40,56 +40,94 @@ public typealias CommunicationSignalingErrorHandler = (CommunicationSignalingErr
...
+    private var configApiVersion: String
 
+    // Step 1: Initialize with basic properties
     init(
         communicationSkypeTokenProvider: CommunicationSkypeTokenProvider,
         logger: ClientLogger = ClientLoggers.default(tag: "AzureCommunicationSignalingClient")
     ) throws {
         self.logger = logger
         self.communicationSkypeTokenProvider = communicationSkypeTokenProvider
+        self.configApiVersion = "2024-09-01"
```
Comment (tjprescott) [inline sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift:59]: Why is this being hard coded?

File: sdk/communication/AzureCommunicationChat/Source/Signaling/RealTimeNotificationConfiguration.swift [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+//
+// --------------------------------------------------------------------------
+
+import Foundation
+
+public struct RealTimeNotificationConfiguration: Codable {
+    public let trouterServiceUrl: String
+    public let registrarServiceUrl: String
+    public let cloudType: String
+}
```
Comment (tjprescott) [inline sdk/communication/AzureCommunicationChat/Source/Signaling/RealTimeNotificationConfiguration.swift:33]: Why is this public? Where's the usage sample for this scenario?

File: sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift (hunk 4 above)
Comment (angellan-msft) [reply sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift:59]: Currently, the service side supports the version "2024-09-01", which is why it's hardcoded in the client. If a new version is released on the service side in the future, we can update this value accordingly on the client side to align with the latest version. Feel free to let me know if you have other opinions!

File: sdk/communication/AzureCommunicationChat/Source/Signaling/RealTimeNotificationConfiguration.swift (hunk 5 above)
Comment (angellan-msft) [reply sdk/communication/AzureCommunicationChat/Source/Signaling/RealTimeNotificationConfiguration.swift]: I updated this class to an internal one.

File: sdk/communication/AzureCommunicationChat/AzureCommunicationChat.podspec [hunk 6]
```diff
@@ -0,0 +1,21 @@
+Pod::Spec.new do |spec|
```
[the remark this answers is not in the document]
Comment (angellan-msft) [reply sdk/communication/AzureCommunicationChat/AzureCommunicationChat.podspec]: I removed this file. (Thanks for pointing it out. This was used for testing and shouldn't be checked in.)

File: sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift [hunk 7]
```diff
@@ -40,56 +40,94 @@ public typealias CommunicationSignalingErrorHandler = (CommunicationSignalingErr
 /// TrouterTokenRefreshHandler for fetching tokens.
 internal typealias TrouterTokenRefreshHandler = (_ stopSignalingClient: Bool, Error?) -> Void
 
-class CommunicationSignalingClient {
-    private var selfHostedTrouterClient: SelfHostedTrouterClient
+public class CommunicationSignalingClient {
```
[the remark this answers is not in the document]
Comment (angellan-msft) [reply sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift]: I updated it back to internal class.

**Review by tjprescott — APPROVED** [review]

**Review by tophpalmer — APPROVED** [review]

**Pull request merged**